### DISCUSSION TITLE
What does Python compatibility mean?

### DISCUSSION THREAD

**Discussion by joshuataylor** [mitsuhiko/minijinja#106] [General]
Hi!

Thought I would turn this into a discussion instead of an issue.

In the compatibility README, it mentions that Python methods are not supported.

Does this include slicing?

I know this is a filter, but doing string and list splitting like this (copied from my test):
```jinja
{
  "foo": "abcdefghijklmnopqrstuvwxyz",
  "test_var": "foo"
}
---
valid strings
:-2 is {{ foo[:-2] }}
:5 is {{ foo[:5] }}
2: is {{ foo[2:] }}
-3: is {{ foo[-3:] }}
3:5 is {{ foo[3:5] }}

valid lists
:1 is {{ [1,2,3][:1] }}
1:2 is {{ [1,2,3][1:2] }}
-2: is {{ [1,2,3][:-2] }}
-2: is {{ [1,2,3][2:] }}
-2: is {{ [1,2,3][-1:-2] }}
-1:3 is {{ [1,2,3][-1:3] }}
-1:-2 is {{ [1,2,3][-1:-2] }}

slicing variables
{{ foo[(test_var|length):] }}
{{ foo[1:(test_var|length)] }}
{{ foo[(test_var|length):-1] }}
{{ foo[(test_var|length):5] }}
```

Is super handy, and seems to be widely used.

What are your thoughts around this? This seems a bit messy to implement in the parser, where in In `parse_postfix()`, in `Token::BracketOpen`:

We have to...
1. check if the first character is a Colon, because we can have values like `:-32`
2. Check if the first value is an int, because it might be a selector and not a slice. Then we parse the expression, then check if the current stream item is a BracketClose, if so it's a `GetItem`. If not, and the next character is a colon,  we check if the next value is NOT a bracket end, if it is we know we can use `parse_expr` to get the second value.
3. If the first value is not an int, just parse the expression, as it could be a variable, or a filter, or whatever. 
4. If the next character is a Colon, progress the stream then check if the current item is a bracket close, if it is then just return. If not, parse the item as the second expression.
4a. If it's not a colon, parse as a GetItem

The compiler for `Expr::Slice` is pretty straight forward, we just check if it has a first/second item and compile each expr. Then compile the general expression after that.

VM becomes a bit more tricky, as we want to support both `ValueRepr::String` and `ValueRepr::Seq`. We need to cover a few edge cases here, but it's pretty straight forward.

**Comment by mitsuhiko** [comment]
Supporting slicing should be trivial to add and I think it’s within the scope of this. I would try to mirror the parsing logic of Jinja2 as close as possible. The parser is quite similar in design between the two.

What I would do however is implement it slightly differently than in Jinja in terms of what it expands to. I would create a value out of the slice expression (eg: a `Slice` value) and then funnel this through `get_item`.

**Comment by mitsuhiko** [comment]
I changed my opinion on the way slices should be implemented to make the object model simpler. You can only slice lists and strings but for that it doesn’t require the value to gain a new primitive.

https://github.com/mitsuhiko/minijinja/pull/120
